feat: native auth on Kubernetes

sparkoo opened the pull request:
Signed-off-by: Michal Vala <mvala@redhat.com>

<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests
-->

### What does this PR do?
 - removed `che.infra.kubernetes.enable_unsupported_k8s` + any checks to it
 - Implemented `OIDCTokenInitializationFilter` that checks token from Authorization header, decodes it and creates Che user from token claims.
 - Moved multiple `Keycloak*` classes into new OIDC module and keep generic OIDC stuff there. Where needed, Keycloak implementation extends these OIDC ones.
 - Moved namespace configuration into `NamespaceConfigurator`s. This makes `KubernetesNamespaceFactory` and `OpenShiftProjectFactory` simpler.
 - Move calling of `NamespaceConfigurator` from `NamespaceProvisioner` into `KubernetesNamespaceFactory`/`OpenShiftProjectFactory`. This makes factories slightly more complex, but still better than doing all configuration/provisioning directly by Factory. 


### Screenshot/screencast of this PR
<!-- Please include a screenshot or a screencast explaining what is doing this PR -->


### What issues does this PR fix or reference?
<!-- Please include any related issue from eclipse che repository (or from another issue tracker).
     Include link to other pull requests like documentation PR from [the docs repo](https://github.com/eclipse/che-docs)
-->
https://github.com/eclipse/che/issues/20633


### How to test this PR?
<!-- Please explain for example :
  - The test platform (openshift, kubernetes, minikube, CodeReady Container, docker-desktop, etc)
  - Installation method: chectl / che-operator
  - steps to reproduce
 -->
che-server image: `quay.io/mvala/che-server:gh20633-nativeAuthKube`
che-operator image: `quay.io/mvala/che-operator:gh20635-nativeAuthKube`

1. start minikube
2. create CheCluster `patch.yaml`, replace `{{MINIKUBE_IP}}` with your minikube IP:
  ```
spec:
    auth:
      nativeUserMode: true
      identityProviderURL: http://dex.dex:5556
      gatewayAuthenticationSidecarImage: quay.io/oauth2-proxy/oauth2-proxy:v7.2.0
      gatewayAuthorizationSidecarImage: quay.io/brancz/kube-rbac-proxy:v0.11.0
    server:
      cheHost: 'che-eclipse-che.{{MINIKUBE_IP}}.nip.io'
      customCheProperties:
        CHE_KEYCLOAK_AUTH__INTERNAL__SERVER__URL: 'http://dex.dex:5556'
```
3. deploy with devworkspaces engine and images (you'll have to accept the risk that chectl prompts) ^^ 
  `chectl server:deploy --platform=minikube --installer=operator --cheimage=quay.io/mvala/che-server:gh20633-nativeAuthKube --che-operator-image=quay.io/mvala/che-operator:gh20635-nativeAuthKube --workspace-engine=dev-workspace --che-operator-cr-patch-yaml=patch.yaml`
4. wait until che is fully deployed (there should be no keycloak), then you can open che url as usual and import the CA cert
5. You get into Dex login page. There are couple of predefined users. `che@eclipse.org:admin` or `userN@che:password` where N is 1-5. After you login, you should get to dashboard as you know.



### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

skabashnyuk commented on pom.xml:
```diff
@@ -1383,6 +1393,12 @@
+                <groupId>io.fabric8</groupId>
+                <artifactId>openshift-server-mock</artifactId>
+                <version>${io.fabric8.kubernetes-client}</version>
+                <scope>test</scope>
```
scope is not needed here

skabashnyuk commented on infrastructures/kubernetes/pom.xml:
```diff
@@ -220,6 +220,11 @@
         </dependency>
+        <dependency>
+            <groupId>com.squareup.okhttp3</groupId>
+            <artifactId>mockwebserver</artifactId>
```
where did it use?

skabashnyuk commented:
@sparkoo I got this error on step3
![Знімок екрана 2021-11-18 о 09 07 23](https://user-images.githubusercontent.com/1614429/142368586-7b5bf4f4-ae4e-4cf9-a6f7-0e1f2a78bc83.png)
my patch.yaml looks like this 

```
spec:
auth:
  nativeUserMode: true
  identityProviderURL: http://dex.dex:5556
server:
  dashboardImage: quay.io/mvala/che-dashboard:kubeNativeAuth
  cheHost: 'che-eclipse-che.192.168.64.7.nip.io'
  customCheProperties:
    CHE_KEYCLOAK_AUTH__INTERNAL__SERVER__URL: 'http://dex.dex:5556'
    CHE_OIDC_USERNAME__CLAIM: 'name
```
```
kubectl get checluster -n eclipse-che -o yaml
apiVersion: v1
items:
- apiVersion: org.eclipse.che/v1
  kind: CheCluster
  metadata:
    creationTimestamp: "2021-11-18T07:06:51Z"
    generation: 1
    name: eclipse-che
    namespace: eclipse-che
    resourceVersion: "960"
    uid: 0d795c86-5930-4d14-a834-f92d0d7c40f6
  status:
    cheClusterRunning: ""
    cheURL: ""
    cheVersion: ""
    dbProvisioned: false
    devfileRegistryURL: ""
    devworkspaceStatus:
      message: DevWorkspace CRDs are not installed
      phase: Inactive
    gitHubOAuthProvisioned: false
    keycloakProvisioned: false
    keycloakURL: ""
    message: Required field "spec.K8s.IngressDomain" is not set
    openShiftoAuthProvisioned: false
    pluginRegistryURL: ""
    reason: InstallOrUpdateFailed
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

skabashnyuk commented on assembly/assembly-wsmaster-war/pom.xml:
```diff
@@ -382,6 +386,11 @@
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-impl</artifactId>
+            <scope>runtime</scope>
```
why did you added `  <scope>runtime</scope>` ?

skabashnyuk replied:
I would add this dependency to exclusion https://github.com/eclipse-che/che-server/blob/672ae77018ace080da9e388ba4a3491e910bcde6/assembly/assembly-wsmaster-war/pom.xml#L435-L457

sparkoo replied:
fixed

skabashnyuk commented on assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:
```diff
@@ -342,6 +342,9 @@ che.infra.kubernetes.service_account_name=NULL
 che.infra.kubernetes.workspace_sa_cluster_roles=NULL
 
+# Cluster roles to assign to user in his namespace
+che.infra.kubernetes.user_cluster_roles=NULL
```
Did we have a similar thing in che-7 workspaces, if not why do we need it in devworkspaces?

skabashnyuk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/CredentialsSecretConfigurator.java:
```diff
@@ -0,0 +1,52 @@
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.NamespaceResolutionContext;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+
```
describe the purpose of this class please

skabashnyuk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/UserPermissionConfigurator.java:
```diff
@@ -0,0 +1,84 @@
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.workspace.infrastructure.kubernetes.CheServerKubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+
```
describe the purpose of this class

skabashnyuk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/WorkspaceServiceAccountConfigurator.java:
```diff
@@ -0,0 +1,68 @@
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesWorkspaceServiceAccount;
+
```
describe the purpose of this class please

skabashnyuk commented on infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/configurator/OpenShiftCredentialsSecretConfigurator.java:
```diff
@@ -0,0 +1,53 @@
+import org.eclipse.che.api.workspace.server.spi.NamespaceResolutionContext;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.configurator.NamespaceConfigurator;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+
```
describe the purpose of this class please

sparkoo replied:
fixed

skabashnyuk commented on infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/configurator/OpenShiftWorkspaceServiceAccountConfigurator.java:
```diff
@@ -0,0 +1,68 @@
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.configurator.NamespaceConfigurator;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+import org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftWorkspaceServiceAccount;
+
```
describe the purpose of this class please

sparkoo replied:
fixed

skabashnyuk commented on infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/configurator/OpenShiftWorkspaceServiceAccountConfigurator.java:
```diff
@@ -0,0 +1,68 @@
+  private final Set<String> clusterRoleNames;
+
+  @Inject
+  public OpenShiftWorkspaceServiceAccountConfigurator(
```
is it somehow different from WorkspaceServiceAccountConfigurator?

sparkoo replied:
this one is using `OpenShiftWorkspaceServiceAccount`, the `WorkspaceServiceAccountConfigurator` is using `KubernetesWorkspaceServiceAccount`.

While it could be done in more generic way, I'm not sure it's worth it. WDYT?

skabashnyuk replied:
I'm fine with that.

skabashnyuk commented on multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakProfileRetriever.java:
```diff
@@ -21,6 +21,7 @@
 import org.eclipse.che.api.core.ApiException;
 import org.eclipse.che.api.core.ServerException;
 import org.eclipse.che.api.core.rest.HttpJsonRequestFactory;
+import org.eclipse.che.multiuser.oidc.OIDCInfo;
```
please check if this import is needed

skabashnyuk commented on multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java:
```diff
@@ -52,6 +52,7 @@
 import org.eclipse.che.dto.server.DtoFactory;
 import org.eclipse.che.multiuser.keycloak.shared.dto.KeycloakErrorResponse;
 import org.eclipse.che.multiuser.keycloak.shared.dto.KeycloakTokenResponse;
+import org.eclipse.che.multiuser.oidc.OIDCInfo;
```
please check if this import is needed

sparkoo replied:
it's constructor parameter

skabashnyuk commented on multiuser/oidc/pom.xml:
```diff
@@ -0,0 +1,87 @@
+        </dependency>
+        <dependency>
+            <groupId>jakarta.servlet</groupId>
+            <artifactId>jakarta.servlet-api</artifactId>
```
please add      <scope>provided</scope>

skabashnyuk commented on multiuser/oidc/src/main/java/org/eclipse/che/multiuser/oidc/filter/OidcTokenInitializationFilter.java:
```diff
@@ -0,0 +1,89 @@
+import org.eclipse.che.multiuser.api.authentication.commons.token.RequestTokenExtractor;
+import org.eclipse.che.multiuser.api.permission.server.AuthorizedSubject;
+import org.eclipse.che.multiuser.api.permission.server.PermissionChecker;
+
```
Please describe the purpose of this class

sparkoo replied:
fixed

mshaposhnik commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/WorkspaceServiceAccountConfigurator.java:
```diff
@@ -0,0 +1,72 @@
+ * This {@link NamespaceConfigurator} ensures that workspace ServiceAccount with proper ClusterRole
+ * is set in Workspace namespace.
+ */
+public class WorkspaceServiceAccountConfigurator implements NamespaceConfigurator {
```
Shouldn't all those configurators to be a `@Singleton`-s ?

sparkoo replied:
fixed

mshaposhnik commented on multiuser/oidc/src/main/java/org/eclipse/che/multiuser/oidc/OIDCSigningKeyResolver.java:
```diff
@@ -0,0 +1,64 @@
+
+  private synchronized PublicKey getJwtPublicKey(JwsHeader<?> header) {
+    String kid = header.getKeyId();
+    if (header.getKeyId() == null) {
```
you may use variable there

sparkoo commented:
@skabashnyuk I've updated che-server and che-operator images. I've also simplified steps to test. Can you please try again and make sure you have latest chectl. The error from your image does not seem related

skabashnyuk commented:
I was able to successfully start devworkspaces.
<img width="1293" alt="Знімок екрана 2021-11-22 о 09 44 57" src="https://user-images.githubusercontent.com/1614429/142827669-bef71045-0a3c-40dc-a466-3175d9aa9ca2.png">

The errors/issues I've got during testing
1. Ingress issue  (reproduced every time)
![Знімок екрана 2021-11-22 о 10 04 38](https://user-images.githubusercontent.com/1614429/142827773-61fd5593-3c6f-4687-a0cb-457e65ba944e.png)
2.Error during first login  (reproduced every time)
<img width="650" alt="Знімок екрана 2021-11-22 о 10 19 01" src="https://user-images.githubusercontent.com/1614429/142827805-63870e85-ffac-4109-8c05-fa658a5a5570.png">
3. Error after logout (reproduced once)
<img width="844" alt="Знімок екрана 2021-11-22 о 09 46 51" src="https://user-images.githubusercontent.com/1614429/142827934-4db13c3b-e1ee-4e95-9da9-d76850ff3bbe.png">

sparkoo commented:
@skabashnyuk 
1. I've never seen that. Looks like it's working afterwards. Looks like chectl issue. Are you on latest chectl version?
2. Yes, I'll report that. IMHO it's happening on openshift with current dev versions as well.
3. Hard to say what exactly happen. This would be on oauth2-proxy I believe. It's possible that as you were messing with the setup, wrong cookie appear in your browser or something like that... Were you able to recover from it? Did "sign-in" worked?

sparkoo commented:
@skabashnyuk is there a script to generate https://github.com/eclipse-che/che-server/blob/main/assembly/assembly-wsmaster-war/.deps/prod.md ?

skabashnyuk commented on infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/OpenshiftTokenInitializationFilterTest.java:
```diff
@@ -128,26 +124,27 @@ public void extractSubjectCreatesSubjectWithCurrentlyAuthenticatedUser()
-
-    verify(servletResponse).sendError(eq(401), anyString());
-  }
+  // TODO: move to abstract class test
```
delete?

sparkoo replied:
fixed

skabashnyuk commented:
```
mvn clean install
cd assembly/ws-maser/war
docker run --rm -t -v ~/.m2:/root/.m2   -v ${PWD}/:/workspace/project  quay.io/che-incubator/dash-licenses:next --generate 
```

metlos commented on infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/OpenshiftTokenInitializationFilterTest.java:
```diff
@@ -128,26 +124,27 @@ public void extractSubjectCreatesSubjectWithCurrentlyAuthenticatedUser()
-
-    verify(servletResponse).sendError(eq(401), anyString());
-  }
+  // TODO: move to abstract class test
```
do you plan to do this in this PR or to create an issue for that?

sparkoo replied:
thanks, I'll update the PR. I've overlooked this

sparkoo replied:
fixed

metlos reviewed (approved):
Nicely done. Just 1 more nitpick about a todo in tests but don't consider being blocked by it...
